tests/unit/compiler/venom/test_common_subexpression_elimination.py

```diff
@@ -0,0 +1,535 @@
+import pytest
+
+from tests.venom_utils import assert_ctx_eq, parse_from_basic_block, parse_venom
+from vyper.venom.analysis.analysis import IRAnalysesCache
+from vyper.venom.passes.common_subexpression_elimination import CSE
+
+
+def _check_pre_post(pre: str, post: str):
+    ctx = parse_from_basic_block(pre)
+    for fn in ctx.functions.values():
+        ac = IRAnalysesCache(fn)
+        CSE(ac, fn).run_pass()
+    assert_ctx_eq(ctx, parse_from_basic_block(post))
+
+
+def _check_pre_post_fn(pre: str, post: str):
+    ctx = parse_venom(pre)
+    for fn in ctx.functions.values():
+        ac = IRAnalysesCache(fn)
+        CSE(ac, fn).run_pass()
+    assert_ctx_eq(ctx, parse_venom(post))
+
+
+def _check_no_change(pre: str):
+    _check_pre_post(pre, pre)
+
+
+def _check_no_change_fn(pre: str):
+    _check_pre_post_fn(pre, pre)
+
+
+def test_common_subexpression_elimination():
+    pre = """
+    main:
+        %1 = param
+        %sum1 = add %1, 10
+        %mul1 = mul %sum1, 10
+        %sum2 = add %1, 10
+        %mul2 = mul %sum2, 10
+        return %mul1, %mul2
+    """
+
+    post = """
+    main:
+        %1 = param
+        %sum1 = add %1, 10
+        %mul1 = mul %sum1, 10
+        %sum2 = %sum1
+        %mul2 = %mul1
+        return %mul1, %mul2
+    """
+
+    _check_pre_post(pre, post)
+
+
+def test_common_subexpression_elimination_commutative():
+    pre = """
+    main:
+        %1 = param
+        %sum1 = add %1, 10
+        %mul1 = mul %sum1, 10
+        %sum2 = add 10, %1
+        %mul2 = mul 10, %sum2
```

Comment:
maybe use variables here, since algebraic optimization pass moves literals to the same position